File: oriedita/src/main/java/com/example/oriedita/editor/handler/MouseHandlerAxiom5.java
package com.example.oriedita.editor.handler;

import jakarta.enterprise.context.ApplicationScoped;
import jakarta.inject.Inject;
import org.apache.commons.lang3.tuple.ImmutablePair;
import org.apache.commons.lang3.tuple.Pair;
import oriedita.editor.canvas.MouseMode;
import origami.Epsilon;
import origami.crease_pattern.OritaCalc;
import origami.crease_pattern.elements.LineColor;
import origami.crease_pattern.elements.LineSegment;
import origami.crease_pattern.elements.Point;
import origami.crease_pattern.elements.StraightLine;

@ApplicationScoped
@Handles(MouseMode.AXIOM_5)
public class MouseHandlerAxiom5 extends BaseMouseHandlerInputRestricted{
    @Inject
    public MouseHandlerAxiom5() {
    }

    //マウス操作(マウスを動かしたとき)を行う関数
    public void mouseMoved(Point p0) {
        if (d.getLineStep().isEmpty()) {
            super.mouseMoved(p0);
        }
    }

    @Override
    public void mousePressed(Point p0) {
        Point p = d.getCamera().TV2object(p0);

        // 1. target point
        if(d.getLineStep().isEmpty()){
            Point closestPoint = d.getClosestPoint(p);
            if (p.distance(closestPoint) < d.getSelectionDistance()) {
                d.lineStepAdd(new LineSegment(closestPoint, closestPoint, d.getLineColor()));
                return;
            }
        }

        // 2. target segment
        if(d.getLineStep().size() == 1){
            LineSegment closestLineSegment = new LineSegment(d.getClosestLineSegment(p));

            if (OritaCalc.determineLineSegmentDistance(p, closestLineSegment) < d.getSelectionDistance()) {
                closestLineSegment.setColor(LineColor.GREEN_6);
                d.lineStepAdd(closestLineSegment);
            }
            return;
        }

        // 3. pivot point
        if(d.getLineStep().size() == 2){
            Point closestPoint = d.getClosestPoint(p);
            if (p.distance(closestPoint) < d.getSelectionDistance()
                    && OritaCalc.determineLineSegmentDistance(closestPoint, d.getLineStep().get(0)) > Epsilon.UNKNOWN_1EN7
                    && !(OritaCalc.isPointWithinLineSpan(closestPoint, d.getLineStep().get(1)) && OritaCalc.isPointWithinLineSpan(d.getLineStep().get(0).getA(), d.getLineStep().get(1)))) {
                d.lineStepAdd(new LineSegment(closestPoint, closestPoint, d.getLineColor()));
                return;
            }
        }

        // index 3 and 4 are the purple indicators
        // 4. indicators (case 1) & destination line (case 2)
        if(d.getLineStep().size() == 5){
            if (OritaCalc.determineLineSegmentDistance(p, d.getLineStep().get(3)) < d.getSelectionDistance() ||
                    OritaCalc.determineLineSegmentDistance(p, d.getLineStep().get(4)) < d.getSelectionDistance()) {
                LineSegment s = d.getClosestLineStepSegment(p, 4, 5);
                s = new LineSegment(s.getB(), s.getA(), d.getLineColor());
                s = (OritaCalc.fullExtendUntilHit(d.getFoldLineSet(), s));

                d.addLineSegment(s);
                d.record();
                d.getLineStep().clear();
                return;
            }

            LineSegment closestLineSegment = new LineSegment(d.getClosestLineSegment(p));

            if (!(OritaCalc.determineLineSegmentDistance(p, closestLineSegment) < d.getSelectionDistance())) {
                return;
            }

            closestLineSegment.setColor(LineColor.GREEN_6);
            d.lineStepAdd(closestLineSegment);
        }
    }

    @Override
    public void mouseDragged(Point p0) {}

    @Override
    public void mouseReleased(Point p0) {
        Point p = d.getCamera().TV2object(p0);

        // first 3 are clicked
        if(d.getLineStep().size() == 3){
            double radius = OritaCalc.distance(d.getLineStep().get(0).getA(), d.getLineStep().get(2).getA());
            drawAxiom5FoldIndicators(radius, d.getLineStep().get(0).getA(), d.getLineStep().get(1), d.getLineStep().get(2).getA());
        }

        // Case 2: Click on destination line and check of mouse point is closer to one of the 2 purple indicators
        if(d.getLineStep().size() == 6){
            Point intersectPoint1 = OritaCalc.findIntersection(d.getLineStep().get(3), d.getLineStep().get(5));
            Point intersectPoint2 = OritaCalc.findIntersection(d.getLineStep().get(4), d.getLineStep().get(5));

            double d1 = OritaCalc.distance(p, intersectPoint1);
            double d2 = OritaCalc.distance(p, intersectPoint2);
            d.addLineSegment(new LineSegment(d.getLineStep().get(2).getA(), d1 < d2 ? intersectPoint1 : intersectPoint2, d.getLineColor()));
            d.record();
            d.getLineStep().clear();
        }
    }
    public void drawAxiom5FoldIndicators(double radius, Point target, LineSegment targetSegment, Point pivot) {
        // Make sure circle radius is not 0
        if(radius <= Epsilon.UNKNOWN_1EN7){ return; }

        double length_a = 0.0; //Distance between the center of the circle and target segment
        Point center = new Point(pivot);

        // If pivot point is not within the target segment span
        if(!OritaCalc.isPointWithinLineSpan(pivot, targetSegment)){
            length_a = OritaCalc.distance(center, OritaCalc.findProjection(targetSegment, center));
        }

        // Intersect at one point
        if(Math.abs(length_a - radius) < Epsilon.UNKNOWN_1EN7){
            Point projectionPoint = OritaCalc.findProjection(targetSegment, pivot);
            LineSegment projectionLine = new LineSegment(pivot, projectionPoint);

            if(OritaCalc.isPointWithinLineSpan(target, projectionLine)){
                if(OritaCalc.distance(projectionPoint, target) < Epsilon.UNKNOWN_1EN7){
                    Point midPoint = new Point(OritaCalc.midPoint(pivot, projectionPoint));

                    d.lineStepAdd(OritaCalc.fullExtendUntilHit(d.getFoldLineSet(), new LineSegment(midPoint, OritaCalc.findProjection(OritaCalc.moveParallel(projectionLine, -1.0), midPoint), LineColor.PURPLE_8)));
                    d.lineStepAdd(OritaCalc.fullExtendUntilHit(d.getFoldLineSet(), new LineSegment(midPoint, OritaCalc.findProjection(OritaCalc.moveParallel(projectionLine, 1.0), midPoint), LineColor.PURPLE_8)));
                    return;
                }

                d.lineStepAdd(OritaCalc.fullExtendUntilHit(d.getFoldLineSet(), new LineSegment(pivot, OritaCalc.findProjection(OritaCalc.moveParallel(projectionLine, 1.0), pivot), LineColor.PURPLE_8)));
                d.lineStepAdd(OritaCalc.fullExtendUntilHit(d.getFoldLineSet(), new LineSegment(pivot, OritaCalc.findProjection(OritaCalc.moveParallel(projectionLine, -1.0), pivot), LineColor.PURPLE_8)));
                return;
            }

            LineSegment s;

            if(OritaCalc.isLineSegmentParallel(new LineSegment(pivot, target), projectionLine) == OritaCalc.ParallelJudgement.NOT_PARALLEL){
                s = OritaCalc.fullExtendUntilHit(d.getFoldLineSet(), new LineSegment(pivot, OritaCalc.center(pivot, target, projectionPoint), LineColor.PURPLE_8));
            } else{
                s = OritaCalc.fullExtendUntilHit(d.getFoldLineSet(), new LineSegment(pivot, projectionPoint, LineColor.PURPLE_8));
            }

            d.lineStepAdd(s);
            d.lineStepAdd(s);
        } else if (length_a > radius) { // Doesn't intersect
            d.getLineStep().clear();
        } else {  // Intersect at two points
            LineSegment l = new LineSegment(target, pivot);
            Point projectPoint = OritaCalc.findProjection(targetSegment, pivot);

            // Length of the last segment of a right triangle (circle center, projection point, and the ultimate guiding point for indicators)
            double length_b = Math.sqrt((radius * radius) - (length_a * length_a));
            LineSegment l1 = processProjectedLineOfIndicator(pivot, projectPoint, length_b);
            LineSegment l2 = processProjectedLineOfIndicator(pivot, projectPoint, -length_b);

            // Recalibrate l1 and l2 for later calculations
            Pair<LineSegment, LineSegment> ls = processPivotWithinSegmentSpan(l1, l2, targetSegment, pivot);
            l1 = ls.getLeft();
            l2 = ls.getRight();

            // Center points for placeholders to draw bisecting indicators on
            Point center1 = processCenter(pivot, l, l1);
            Point center2 = processCenter(pivot, l, l2);

            // Decide the indicators on different cases
            determineIndicators(l, l1, l2, pivot, center1, center2, target, targetSegment, pivot);
        }
    }

    private LineSegment processProjectedLineOfIndicator(Point pivot, Point projectPoint, double length){
        LineSegment projectLine = new LineSegment(pivot, projectPoint);
        LineSegment line = new LineSegment(projectPoint, OritaCalc.findProjection(OritaCalc.moveParallel(projectLine, length), projectPoint));
        return new LineSegment(pivot, line.getB());
    }

    private Point processCenter(Point pivot, LineSegment l1, LineSegment l2){
        // l1 and l2 are 2 lines forming a triangle with a common point
        // If l1 and l2 are aligned/parallel
        //       l1 ⤵              l2 ⤵
        // ------------------O-------------------
        //     common point ⤴
        if(OritaCalc.isLineSegmentParallel(new StraightLine(l1.determineFurthestEndpoint(pivot), pivot), new StraightLine(pivot, l2.determineFurthestEndpoint(pivot))) == OritaCalc.ParallelJudgement.PARALLEL_EQUAL){
            LineSegment seg = new LineSegment(pivot, OritaCalc.findProjection(OritaCalc.moveParallel(l1, 1), pivot));
            // return the center using a different point on a shifted line parallel to l2
            return OritaCalc.center(l1.determineFurthestEndpoint(pivot), l2.determineFurthestEndpoint(pivot), seg.determineFurthestEndpoint(pivot));
        } else { // return the default center
            return OritaCalc.center(pivot, l2.determineFurthestEndpoint(pivot), l1.determineFurthestEndpoint(pivot));
        }
    }

    private Pair<LineSegment, LineSegment> processPivotWithinSegmentSpan(LineSegment l1, LineSegment l2, LineSegment targetSegment, Point pivot){
        // If pivot point is within the target segment span
        if(OritaCalc.isPointWithinLineSpan(pivot, targetSegment)){
            // pivot within span
            if(OritaCalc.distance(pivot, targetSegment.getA()) < Epsilon.UNKNOWN_1EN7){
                //pivot within span touching A
                l1 = (new LineSegment(pivot, OritaCalc.point_rotate(pivot, targetSegment.getB(), 180)));
                l2 = (new LineSegment(pivot, targetSegment.getB()));
                return new ImmutablePair<>(l1, l2);
            }
            if(OritaCalc.distance(pivot, targetSegment.getB()) < Epsilon.UNKNOWN_1EN7){
                //pivot within span touching B
                l1 = (new LineSegment(pivot, targetSegment.getA()));
                l2 = (new LineSegment(pivot, OritaCalc.point_rotate(pivot, targetSegment.getA(), 180)));
                return new ImmutablePair<>(l1, l2);
            }

            boolean isOutsideSegmentA = targetSegment.determineLength() > OritaCalc.distance(targetSegment.getA(), pivot) &&
                    OritaCalc.distance(targetSegment.getB(), pivot) > targetSegment.determineLength();
            boolean isOutsideSegmentB = targetSegment.determineLength() > OritaCalc.distance(targetSegment.getB(), pivot) &&
                    OritaCalc.distance(targetSegment.getA(), pivot) > targetSegment.determineLength();

            // If pivot point is outside the target segment
            //pivot within span outside A
            l1 = (new LineSegment(pivot,isOutsideSegmentA ? OritaCalc.point_rotate(pivot, targetSegment.getB(), 180) : targetSegment.getA()));

            // pivot within span outside B
            l2 = (new LineSegment(pivot,isOutsideSegmentB ? OritaCalc.point_rotate(pivot, targetSegment.getA(), 180) : targetSegment.getB()));
            return new ImmutablePair<>(l1, l2);
        }
        return new ImmutablePair<>(l1, l2);
    }

    private void determineIndicators(LineSegment l, LineSegment l1, LineSegment l2, Point center, Point center1, Point center2, Point target, LineSegment targetSegment, Point pivot){
        if(OritaCalc.distance(center1, OritaCalc.findProjection(targetSegment, center1)) > Epsilon.UNKNOWN_1EN7 ||
                OritaCalc.distance(center2, OritaCalc.findProjection(targetSegment, center2)) > Epsilon.UNKNOWN_1EN7){
            // If target point is not within target segment span
            if(!OritaCalc.isPointWithinLineSpan(target, targetSegment)){
                //target not in span 1
                LineSegment s1 = OritaCalc.fullExtendUntilHit(d.getFoldLineSet(), new LineSegment(center, center1, LineColor.PURPLE_8));
                LineSegment s2 = OritaCalc.fullExtendUntilHit(d.getFoldLineSet(), new LineSegment(center, center2, LineColor.PURPLE_8));

                d.lineStepAdd(s1);
                d.lineStepAdd(s2);
                return;
            }
            // If target point is within target segment span
            if(OritaCalc.isLineSegmentParallel(l1, l) == OritaCalc.ParallelJudgement.PARALLEL_EQUAL){
                //target not in span 2
                LineSegment s = OritaCalc.fullExtendUntilHit(d.getFoldLineSet(), new LineSegment(center, center2, LineColor.PURPLE_8));

                d.lineStepAdd(s);
                d.lineStepAdd(s);
                return;
            }
            if(OritaCalc.isLineSegmentParallel(l2, l) == OritaCalc.ParallelJudgement.PARALLEL_EQUAL){
                //target not in span 3
                LineSegment s = OritaCalc.fullExtendUntilHit(d.getFoldLineSet(), new LineSegment(center, center1, LineColor.PURPLE_8));

                d.lineStepAdd(s);
                d.lineStepAdd(s);
            }
        } else{
            LineSegment s1 = OritaCalc.fullExtendUntilHit(d.getFoldLineSet(), new LineSegment(pivot, OritaCalc.findProjection(OritaCalc.moveParallel(l1, 1.0), pivot), LineColor.PURPLE_8));
            LineSegment s2 = OritaCalc.fullExtendUntilHit(d.getFoldLineSet(), new LineSegment(pivot, OritaCalc.findProjection(OritaCalc.moveParallel(l2, -1.0), pivot), LineColor.PURPLE_8));

            d.lineStepAdd(s1);
            d.lineStepAdd(s2);
        }
    }
}

File: oriedita-ui/src/main/java/com/example/oriedita_ui/handler/MouseHandlerAxiom5.kt
package com.example.oriedita_ui.handler

import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.input.pointer.PointerInputChange
import com.example.oriedita_core.origami.crease_pattern.elements.LineSegment
import com.example.oriedita_core.origami.crease_pattern.elements.Point
import com.example.oriedita_core.origami.crease_pattern.elements.LineColor
import com.example.oriedita_core.origami.crease_pattern.OritaCalc
import kotlin.math.abs
import kotlin.math.sqrt

/**
 * Обработчик аксиомы 5 оригами
 * Адаптированная версия для Android
 */
class MouseHandlerAxiom5 : BaseMouseHandler() {
    
    private var targetPoint: Point? = null
    private var targetSegment: LineSegment? = null
    private var pivotPoint: Point? = null
    private var indicator1: LineSegment? = null
    private var indicator2: LineSegment? = null
    private var destinationSegment: LineSegment? = null
    private var currentStep = Axiom5Step.SELECT_TARGET_POINT
    private var isProcessing = false
    
    override fun onPress(offset: Offset, change: PointerInputChange): Boolean {
        val point = offsetToPoint(offset)
        
        when (currentStep) {
            Axiom5Step.SELECT_TARGET_POINT -> {
                handleSelectTargetPointPress(point)
            }
            Axiom5Step.SELECT_TARGET_SEGMENT -> {
                handleSelectTargetSegmentPress(point)
            }
            Axiom5Step.SELECT_PIVOT_POINT -> {
                handleSelectPivotPointPress(point)
            }
            Axiom5Step.SELECT_DESTINATION_OR_INDICATOR -> {
                handleSelectDestinationOrIndicatorPress(point)
            }
        }
        
        return true
    }
    
    override fun onDrag(offset: Offset, change: PointerInputChange): Boolean {
        val point = offsetToPoint(offset)
        
        when (currentStep) {
            Axiom5Step.SELECT_TARGET_POINT -> {
                handleSelectTargetPointDrag(point)
            }
            Axiom5Step.SELECT_TARGET_SEGMENT -> {
                handleSelectTargetSegmentDrag(point)
            }
            Axiom5Step.SELECT_PIVOT_POINT -> {
                handleSelectPivotPointDrag(point)
            }
            Axiom5Step.SELECT_DESTINATION_OR_INDICATOR -> {
                handleSelectDestinationOrIndicatorDrag(point)
            }
        }
        
        return true
    }
    
    override fun onRelease(offset: Offset, change: PointerInputChange): Boolean {
        val point = offsetToPoint(offset)
        
        when (currentStep) {
            Axiom5Step.SELECT_TARGET_POINT -> {
                currentStep = handleSelectTargetPointRelease(point)
            }
            Axiom5Step.SELECT_TARGET_SEGMENT -> {
                currentStep = handleSelectTargetSegmentRelease(point)
            }
            Axiom5Step.SELECT_PIVOT_POINT -> {
                currentStep = handleSelectPivotPointRelease(point)
            }
            Axiom5Step.SELECT_DESTINATION_OR_INDICATOR -> {
                currentStep = handleSelectDestinationOrIndicatorRelease(point)
            }
        }
        
        return true
    }
    
    private fun handleSelectTargetPointPress(point: Point) {
        val closestPoint = getClosestPoint(point)
        if (point.distance(closestPoint) < getSelectionDistance()) {
            targetPoint = closestPoint
        } else {
            targetPoint = null
        }
        
        println("Целевая точка: ${targetPoint?.let { "(${it.x}, ${it.y})" } ?: "не выбрана"}")
    }
    
    private fun handleSelectTargetPointDrag(point: Point) {
        handleSelectTargetPointPress(point)
    }
    
    private fun handleSelectTargetPointRelease(point: Point): Axiom5Step {
        if (targetPoint == null) return Axiom5Step.SELECT_TARGET_POINT
        
        println("Переход к выбору целевого сегмента")
        return Axiom5Step.SELECT_TARGET_SEGMENT
    }
    
    private fun handleSelectTargetSegmentPress(point: Point) {
        val closestLineSegment = getClosestLineSegment(point)
        if (OritaCalc.determineLineSegmentDistance(point, closestLineSegment) < getSelectionDistance()) {
            targetSegment = LineSegment(closestLineSegment.getA(), closestLineSegment.getB(), LineColor.GREEN_6)
        } else {
            targetSegment = null
        }
    }
    
    private fun handleSelectTargetSegmentDrag(point: Point) {
        handleSelectTargetSegmentPress(point)
    }
    
    private fun handleSelectTargetSegmentRelease(point: Point): Axiom5Step {
        if (targetSegment == null) return Axiom5Step.SELECT_TARGET_SEGMENT
        
        println("Переход к выбору точки поворота")
        return Axiom5Step.SELECT_PIVOT_POINT
    }
    
    private fun handleSelectPivotPointPress(point: Point) {
        val closestPoint = getClosestPoint(point)
        if (point.distance(closestPoint) < getSelectionDistance() &&
            OritaCalc.determineLineSegmentDistance(closestPoint, LineSegment(targetPoint!!, targetPoint!!)) > 0.0000001 &&
            !(OritaCalc.isPointWithinLineSpan(closestPoint, targetSegment!!) && 
              OritaCalc.isPointWithinLineSpan(targetPoint!!, targetSegment!!))) {
            pivotPoint = closestPoint
        } else {
            pivotPoint = null
        }
    }
    
    private fun handleSelectPivotPointDrag(point: Point) {
        handleSelectPivotPointPress(point)
    }
    
    private fun handleSelectPivotPointRelease(point: Point): Axiom5Step {
        if (pivotPoint == null) return Axiom5Step.SELECT_PIVOT_POINT
        
        val radius = OritaCalc.distance(targetPoint!!, pivotPoint!!)
        drawAxiom5FoldIndicators(radius)
        
        println("Переход к выбору назначения или индикатора")
        return Axiom5Step.SELECT_DESTINATION_OR_INDICATOR
    }
    
    private fun handleSelectDestinationOrIndicatorPress(point: Point) {
        updateDestinationSegment(point)
    }
    
    private fun handleSelectDestinationOrIndicatorDrag(point: Point) {
        updateDestinationSegment(point)
    }
    
    private fun handleSelectDestinationOrIndicatorRelease(point: Point): Axiom5Step {
        if (OritaCalc.determineLineSegmentDistance(point, indicator1!!) < getSelectionDistance() ||
            OritaCalc.determineLineSegmentDistance(point, indicator2!!) < getSelectionDistance()) {
            
            val selectedIndicator = if (OritaCalc.determineLineSegmentDistance(point, indicator1!!) < 
                                       OritaCalc.determineLineSegmentDistance(point, indicator2!!)) {
                indicator1!!
            } else {
                indicator2!!
            }
            
            val newSegment = LineSegment(selectedIndicator.getB(), selectedIndicator.getA(), getLineColor())
            val extendedSegment = OritaCalc.fullExtendUntilHit(getFoldLineSet(), newSegment)
            
            addLineSegment(extendedSegment)
            println("Добавлена линия аксиомы 5")
            
            reset()
            return Axiom5Step.SELECT_TARGET_POINT
        }
        
        if (destinationSegment == null) return Axiom5Step.SELECT_DESTINATION_OR_INDICATOR
        
        val intersectPoint1 = OritaCalc.findIntersection(indicator1!!, destinationSegment!!)
        val intersectPoint2 = OritaCalc.findIntersection(indicator2!!, destinationSegment!!)
        
        val d1 = OritaCalc.distance(point, intersectPoint1)
        val d2 = OritaCalc.distance(point, intersectPoint2)
        
        val newSegment = LineSegment(pivotPoint!!, if (d1 < d2) intersectPoint1 else intersectPoint2, getLineColor())
        addLineSegment(newSegment)
        
        println("Добавлена линия аксиомы 5")
        reset()
        return Axiom5Step.SELECT_TARGET_POINT
    }
    
    private fun updateDestinationSegment(point: Point) {
        val indicator1Distance = OritaCalc.determineLineSegmentDistance(point, indicator1!!)
        val indicator2Distance = OritaCalc.determineLineSegmentDistance(point, indicator2!!)
        val targetSegmentDistance = OritaCalc.determineLineSegmentDistance(point, targetSegment!!)
        val normalDistance = OritaCalc.determineLineSegmentDistance(point, getClosestLineSegment(point))
        
        val minDistance = minOf(indicator1Distance, indicator2Distance, targetSegmentDistance, normalDistance)
        
        destinationSegment = when {
            abs(minDistance - indicator1Distance) < 0.000001 && indicator1Distance < getSelectionDistance() -> {
                LineSegment(indicator1!!.getA(), indicator1!!.getB(), LineColor.ORANGE_4)
            }
            abs(minDistance - indicator2Distance) < 0.000001 && indicator2Distance < getSelectionDistance() -> {
                LineSegment(indicator2!!.getA(), indicator2!!.getB(), LineColor.ORANGE_4)
            }
            abs(minDistance - targetSegmentDistance) < 0.000001 && targetSegmentDistance < getSelectionDistance() -> {
                null
            }
            abs(minDistance - normalDistance) < 0.000001 && normalDistance < getSelectionDistance() -> {
                val closestSegment = getClosestLineSegment(point)
                LineSegment(closestSegment.getA(), closestSegment.getB(), LineColor.ORANGE_4)
            }
            else -> null
        }
    }
    
    private fun drawAxiom5FoldIndicators(radius: Double) {
        if (radius <= 0.0000001) {
            reset()
            return
        }
        
        val center = Point(pivotPoint!!)
        val lengthA = if (!OritaCalc.isPointWithinLineSpan(pivotPoint!!, targetSegment!!)) {
            OritaCalc.distance(center, OritaCalc.findProjection(targetSegment!!, center))
        } else {
            0.0
        }
        
        if (abs(lengthA - radius) < 0.0000001) {
            // Пересечение в одной точке
            val projectionPoint = OritaCalc.findProjection(targetSegment!!, pivotPoint!!)
            val projectionLine = LineSegment(pivotPoint!!, projectionPoint)
            
            if (OritaCalc.isPointWithinLineSpan(targetPoint!!, projectionLine)) {
                if (OritaCalc.distance(projectionPoint, targetPoint!!) < 0.0000001) {
                    val midPoint = Point(OritaCalc.midPoint(pivotPoint!!, projectionPoint))
                    
                    indicator1 = OritaCalc.fullExtendUntilHit(getFoldLineSet(), 
                        LineSegment(midPoint, OritaCalc.findProjection(OritaCalc.moveParallel(projectionLine, -1.0), midPoint), LineColor.PURPLE_8))
                    indicator2 = OritaCalc.fullExtendUntilHit(getFoldLineSet(), 
                        LineSegment(midPoint, OritaCalc.findProjection(OritaCalc.moveParallel(projectionLine, 1.0), midPoint), LineColor.PURPLE_8))
                } else {
                    indicator1 = OritaCalc.fullExtendUntilHit(getFoldLineSet(), 
                        LineSegment(pivotPoint!!, OritaCalc.findProjection(OritaCalc.moveParallel(projectionLine, 1.0), pivotPoint!!), LineColor.PURPLE_8))
                    indicator2 = OritaCalc.fullExtendUntilHit(getFoldLineSet(), 
                        LineSegment(pivotPoint!!, OritaCalc.findProjection(OritaCalc.moveParallel(projectionLine, -1.0), pivotPoint!!), LineColor.PURPLE_8))
                }
            } else {
                val s = if (OritaCalc.isLineSegmentParallel(LineSegment(pivotPoint!!, targetPoint!!), projectionLine) == OritaCalc.ParallelJudgement.NOT_PARALLEL) {
                    OritaCalc.fullExtendUntilHit(getFoldLineSet(), 
                        LineSegment(pivotPoint!!, OritaCalc.center(pivotPoint!!, targetPoint!!, projectionPoint), LineColor.PURPLE_8))
                } else {
                    OritaCalc.fullExtendUntilHit(getFoldLineSet(), 
                        LineSegment(pivotPoint!!, projectionPoint, LineColor.PURPLE_8))
                }
                indicator1 = s
                indicator2 = s
            }
        } else if (lengthA > radius) {
            // Не пересекается
            reset()
        } else {
            // Пересечение в двух точках
            val lengthB = sqrt((radius * radius) - (lengthA * lengthA))
            val projectPoint = OritaCalc.findProjection(targetSegment!!, pivotPoint!!)
            
            val l1 = processProjectedLineOfIndicator(pivotPoint!!, projectPoint, lengthB)
            val l2 = processProjectedLineOfIndicator(pivotPoint!!, projectPoint, -lengthB)
            
            val center1 = processCenter(pivotPoint!!, LineSegment(targetPoint!!, pivotPoint!!), l1)
            val center2 = processCenter(pivotPoint!!, LineSegment(targetPoint!!, pivotPoint!!), l2)
            
            indicator1 = OritaCalc.fullExtendUntilHit(getFoldLineSet(), 
                LineSegment(center, center1, LineColor.PURPLE_8))
            indicator2 = OritaCalc.fullExtendUntilHit(getFoldLineSet(), 
                LineSegment(center, center2, LineColor.PURPLE_8))
        }
        
        println("Индикаторы аксиомы 5 созданы")
    }
    
    private fun processProjectedLineOfIndicator(pivot: Point, projectPoint: Point, length: Double): LineSegment {
        val projectLine = LineSegment(pivot, projectPoint)
        val line = LineSegment(projectPoint, OritaCalc.findProjection(OritaCalc.moveParallel(projectLine, length), projectPoint))
        return LineSegment(pivot, line.getB())
    }
    
    private fun processCenter(pivot: Point, l1: LineSegment, l2: LineSegment): Point {
        return OritaCalc.center(pivot, l2.getB(), l1.getB())
    }
    
    override fun reset() {
        targetPoint = null
        targetSegment = null
        pivotPoint = null
        indicator1 = null
        indicator2 = null
        destinationSegment = null
        currentStep = Axiom5Step.SELECT_TARGET_POINT
        isProcessing = false
        
        println("Аксиома 5 сброшена")
    }
    
    fun getCurrentStep(): Axiom5Step = currentStep
    
    fun getTargetPoint(): Point? = targetPoint
    
    fun getTargetSegment(): LineSegment? = targetSegment
    
    fun getPivotPoint(): Point? = pivotPoint
    
    fun getIndicators(): Pair<LineSegment?, LineSegment?> = Pair(indicator1, indicator2)
    
    fun getDestinationSegment(): LineSegment? = destinationSegment
    
    fun isProcessing(): Boolean = isProcessing
    
    fun getAxiom5Description(): String {
        return "Аксиома 5 - шаг: ${currentStep.name}, целевая точка: ${targetPoint?.let { "(${it.x}, ${it.y})" } ?: "не выбрана"}"
    }
    
    override fun getName(): String = "Аксиома 5"
    
    override fun getDescription(): String = "Создавайте складки через точку и линию"
}

enum class Axiom5Step {
    SELECT_TARGET_POINT,
    SELECT_TARGET_SEGMENT,
    SELECT_PIVOT_POINT,
    SELECT_DESTINATION_OR_INDICATOR
} 